plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Cse.hs

```diff
@@ -0,0 +1,398 @@
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TupleSections     #-}
+{-# LANGUAGE TypeApplications  #-}
+{-# LANGUAGE TypeFamilies      #-}
+{-# LANGUAGE TypeOperators     #-}
+
+module UntypedPlutusCore.Transform.Cse (cse) where
+
+import PlutusCore (MonadQuote, Name, freshName)
+import UntypedPlutusCore.Core
+import UntypedPlutusCore.Purity (isWorkFree)
+import UntypedPlutusCore.Size
+
+import Control.Lens
+import Control.Monad
+import Control.Monad.Trans.Class
+import Control.Monad.Trans.Reader
+import Control.Monad.Trans.State.Strict
+import Data.Foldable
+import Data.Hashable
+import Data.HashMap.Strict (HashMap)
+import Data.HashMap.Strict qualified as Map
+import Data.List.Extra
+import Data.Traversable
+
+{- Note [CSE]
+
+-------------------------------------------------------------------------------
+1. Simplifications
+-------------------------------------------------------------------------------
+
+This is a simplified (i.e., not fully optimal) implementation of CSE. The two simplifications
```

Comment:
Did you use any references for the implementation? Is this a standard algorithm? What variants do we know of? What does GHC do?

Reply:
Ok, I'll expand the notes a bit